Sources/Crust/Parser.swift

```diff
@@ -596,26 +615,22 @@ extension Parser {
     }
 
     // Gather all the identifiers.
+    var names = [QualifiedNameSyntax]()
     while case .identifier(_) = peek() {
-      pieces.append(currentToken!)
-      advance()
+      names.append(try parseQualifiedName())
     }
 
     // If we've not hit the matching closing paren, we must be parsing a typed
     // parameter group
     //
     // (a b c ... : <expr>) {d e f ... : <expr>} ...
     if case .colon = peek() {
-      return try self.finishParsingTypedParameterGroupExpr(leftParen, pieces)
+      return try self.finishParsingTypedParameterGroupExpr(leftParen, names)
     }
 
     // Else consume the closing paren.
```

Comment:
Recur here to parse a basic expr if you encounter a left paren.

Reply:
Better idea: parse basic expressions instead of names, then validate that we only got names once we've seen a `:`.